### PR TITLE
Report FQDN

### LINKED ISSUE #149: Report FQDN on HostInfo
Report the FQDN on HostInfo

 - Relates https://github.com/elastic/elastic-agent/issues/2185
 - Relates https://github.com/elastic/beats/issues/1070

### PR DESCRIPTION
**Pull request by AndersonQ**
This PR adds FQDN to `types.HostInfo` and expands the `linux`, `darwin` and `windows` providers to retrieve and populate the newly added fields.

The FQDN is obtained by:

 - Linux/Darwin:
   - `net.LookupCNAME` for hostname
   - reverse lookup by hostname:
     - `net.LookupIP` for hostname
     - `net.LookupAddr` for the obateined IP
 - Windows:
   - get [`ComputerNamePhysicalDnsFullyQualified`]( https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getcomputernameexw#return-value)

### Integration tests:

There are integration tests for linux, which requires docker, to run them use the build tag `integration`. The `TestHost_FQDN` automatically starts docker containers and runs the tests.

```
GO_VERSION=1.19 go test -tags integration -v -count 1 -run TestHost_FQDN$ ./providers/linux/
```

### Manually:

Run the test `TestHost` for the desired provider, `linux`, `windows` or `darwin`:
```
go test -v -run TestHost$ ./providers/PROVIDER/
```

On docker:
```
docker run --rm -v "$PWD":/usr/src/elastic/go-sysinfo -w /usr/src/elastic/go-sysinfo golang:1.19 go test -v -run TestHost$ ./providers/linux 
```

The `TestHost` test only prints the information collected, you need to check it for correctness. 

look for 
```
          "name": "mokona-elastic",
          "fqdn": "mokona-elastic.lan",
```

<details>
  <summary>full output</summary>
  
  ```text
❯ go test -v -run TestHost$ ./providers/linux/   

=== RUN   TestHost
    host_linux_test.go:39: {
          "architecture": "x86_64",
          "boot_time": "2023-01-23T07:04:06+01:00",
          "containerized": false,
          "name": "mokona-elastic",
          "domain": "",
          "fqdn": "mokona-elastic.lan",
          "ip": [
            "127.0.0.1/8",
            "::1/128",
            "192.168.86.136/24",
            "fd66:61eb:174c:0:5961:a7be:b434:e131/64",
            "fd66:61eb:174c:0:95a:f7a1:cec2:6d8a/64",
            "fe80::780d:6555:d649:7934/64",
            "172.26.0.1/16",
            "172.17.0.1/16",
            "fe80::42:61ff:fec1:1585/64",
            "172.18.0.1/16",
            "fc00:f853:ccd:e793::1/64",
            "fe80::1/64",
            "172.25.0.1/16",
            "172.19.0.1/16",
            "172.21.0.1/16",
            "172.20.0.1/16"
          ],
          "kernel_version": "6.0.6-76060006-generic",
          "mac": [
            "00:e0:4c:ad:f7:bb",
            "7c:21:4a:a3:e2:dd",
            "02:42:c8:ee:00:03",
            "02:42:61:c1:15:85",
            "02:42:fd:6b:4f:5d",
            "02:42:e2:b4:8d:10",
            "02:42:61:c8:02:f2",
            "02:42:29:aa:6d:01",
            "02:42:77:1f:61:c9"
          ],
          "os": {
            "type": "linux",
            "family": "",
            "platform": "pop",
            "name": "Pop!_OS",
            "version": "22.04 LTS",
            "major": 22,
            "minor": 4,
            "patch": 0,
            "codename": "jammy"
          },
          "timezone": "CET",
          "timezone_offset_sec": 3600,
          "id": "e9ed195eddb96448172f94386284db62"
        }
--- PASS: TestHost (0.00s)
PASS
ok  	github.com/elastic/go-sysinfo/providers/linux	0.192s
  ```
</details>

Tested on:

 - [x] Linux
   - [x] Native
   - [x] Docker
   - [ ] K8s (should it be an e2e test on the Elastic Agent side?)
 - [ ] Mac
   - [x] Native
   - [x] Docker (is it required?)
 - [x] Windows
 - [ ] ~Aix~

 - closes #149
 - Relates to https://github.com/elastic/beats/issues/1070

### CODE REVIEW

File: providers/linux/host_linux.go [hunk 1]
```diff
@@ -138,7 +141,7 @@ func (h *host) CPUTime() (types.CPUTimes, error) {
 }
 
 func newHost(fs procFS) (*host, error) {
-	stat, err := fs.NewStat()
+	stat, err := fs.Stat()
```
Comment (fearful-symmetry) [inline providers/linux/host_linux.go:144]: I assume you changed this since `NewStat()` is marked as deprecated---in which case, there's a few more instances if that you might want to change

File: providers/linux/host_linux.go [hunk 2]
```diff
@@ -210,6 +215,24 @@ func (r *reader) hostname(h *host) {
 	h.info.Hostname = v
 }
 
+func (r *reader) fqdn(h *host) {
+	const cmdPath, args = "/bin/hostname", "-f"
```
Comment (fearful-symmetry) [inline providers/linux/host_linux.go:219]: Since you asked about this this morning, I assume you know, but beat's security policy usually prevents us from exec'ing out like this.

**Review by fearful-symmetry — COMMENTED** [review]
Just leaving a few drive-by comments. Also, don't forget to update the main feature chart thing in the `readme.md`.

File: providers/linux/host_linux.go (hunk 2 above)
Comment (AndersonQ) [reply providers/linux/host_linux.go:219]: Oh, I didn't know, thanks for the heads up

File: providers/darwin/host_darwin.go [hunk 3]
```diff
@@ -20,11 +20,13 @@
 
 package darwin
 
+import "C"
```
Comment (andrewkroh) [inline providers/darwin/host_darwin.go:23]: I realize this is a draft, but wanted to mention that the providers need to continue to function without CGO even if it is in a degraded manner (like not reporting the FQDN). Checkout https://github.com/elastic/go-sysinfo#goos--goarch-pairs.

**Comment by joshdover** [comment]
@AndersonQ Just to make sure we're on the same page, this PR can still continue even before we make the decision on the default behavior in Beats/Agent.

**Comment by leehinman** [comment]
Results from when I ran on MacOS 12.6.2 arm64 

| Environment   | CGO_ENABLED | name         | fqdn             |
|---------------| ------------|--------------|------------------|
| darwin        | 0           | elastic2     | elastic2         |
| darwin        | 1           | elastic2     | elastic2.lan     |
| darwin docker | 0           | dee3070d2dc3 | dee3070d2dc3     |
| darwin docker | 1           | a0914c8b7bf2 | a0914c8b7bf2.lan |

File: types/host.go [hunk 4]
```diff
@@ -63,17 +63,20 @@ type VMStat interface {
...
-	Timezone          string    `json:"timezone"`                // System timezone.
-	TimezoneOffsetSec int       `json:"timezone_offset_sec"`     // Timezone offset (seconds from UTC).
-	UniqueID          string    `json:"id,omitempty"`            // Unique ID of the host (optional).
+	Architecture  string    `json:"architecture"`            // Hardware architecture (e.g. x86_64, arm, ppc, mips).
+	BootTime      time.Time `json:"boot_time"`               // Host boot time.
+	Containerized *bool     `json:"containerized,omitempty"` // Is the process containerized.
+	Hostname      string    `json:"name"`                    // Hostname
+	// FQDN is the Fully qualified domain name, for linux and darwin it requires CGO.
+	// if CGO is not enabled it'll fallback to os.Hostname(), the same as Hostname.
+	FQDN              string   `json:"fqdn"`
```
Comment (andrewkroh) [inline types/host.go:72]: As a user of the library I would prefer the library indicate when it cannot collect the FQDN rather than silently fallback. IMO the most clear way to accomplish this would be to use `*string` and leave it nil when the data is unavailable. Then document similar to what you have that not a provider implementations support this field and it may be nil.

**Review by andrewkroh — COMMENTED** [review]
Given that FQDNs are case-insensitive I think it would be helpful for this library to normalize the values to lowercase despite what the OS gives. This will help ensure consistency downstream.

File: types/host.go (hunk 4 above)
Comment (AndersonQ) [reply types/host.go:74]: done, I added Domain. Now if Domain cannot be retrieved, Cgo is disabled for linux and darwin, FQDN will be empty as well as Domain.

**Comment by leehinman** [comment]
Re-ran tests on MacOS 12.6.2 arm64

| Environment   | CGO_ENABLED | name           | domain | fqdn               |
|---------------|-------------|----------------|--------|--------------------|
| darwin        | 0           | "elastic2"     | ""     | ""                 |
| darwin        | 1           | "elastic2"     | ""     | "elastic2.lan"     |
| darwin docker | 0           | "5bad5ca5beca" | ""     | ""                 |
| darwin docker | 1           | "96b268cfda81" | ""     | "96b268cfda81.lan" |

**Comment by AndersonQ** [comment]
> Given that FQDNs are case-insensitive I think it would be helpful for this library to normalize the values to lowercase despite what the OS gives.

it does make sense. but then it'd also be true for the hostname and domain, right? For consistency, I think they should be aligned. However it'd be a breaking change for hostname, what I'd say is less than ideal... 

In that sense, also considering it's just a lib, I'd let this decision for whoever is using it. We returns as we get, the user transforms as they need.

What do you think?

**Comment by fearful-symmetry** [comment]
Alright, current round of CI errors seems unrelated?

**Comment by AndersonQ** [comment]
> Alright, current round of CI errors seems unrelated?

@fearful-symmetry, @andrewkroh do any of you are familiar with these tests "failing"?

The culprit is `Archive JUnit-formatted test results`, but [its output](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp/detail/PR-144/37/pipeline/154) is the same as for others that are green

```
[2023-01-24T16:45:09.128Z] Recording test results

[2023-01-24T16:45:11.474Z] [Checks API] No suitable checks publisher found.
```

**Comment by andrewkroh** [comment]
> In that sense, also considering it's just a lib, I'd let this decision for whoever is using it. We returns as we get, the user transforms as they need.

That's fine. Either when we use this in Beats / Agent or when we create the mappings for this field we should ensure that the value is normalized.

**Comment by andrewkroh** [comment]
> do any of you are familiar with these tests "failing"?

Some of the FQDN tests are failing on ubuntu under both CGO_ENABLED=0 and CGO_ENABLED=1. See the "Tests" tab:

https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp/detail/PR-144/37/tests

File: providers/darwin/host_fqdn_cgo_darwin.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+// #include <stdlib.h>
+import "C"
+import (
+	"fmt"
+	"os"
+	"unsafe"
+)
+
+func fqdn() (string, error) {
+	hostname, err := os.Hostname()
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply providers/darwin/host_fqdn_cgo_darwin.go:32]: It gets even more confusing if the machine has a NIS domain set AND the HostName is fully-qualified. I think we need to be clearify what we are reporting. Do we want to report the NIS domain or do we want to report domain suffix that is part of the machine's fully-qualified DNS name.

`sudo sysctl kern.nisdomainname=nis.domain`

```
            "name": "foo.engineering.example.com",
            "domain": "nis.domain",
            "fqdn": "foo.engineering.example.com.nis.domain",
```

File: providers/darwin/process_darwin_test.go [hunk 6]
```diff
@@ -104,39 +104,28 @@ func TestProcesses(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	var count int
```
Comment (andrewkroh) [inline providers/darwin/process_darwin_test.go:107]: Could you revert these changes now that #150 is in.

File: providers/linux/host_fqdn_nocgo_linux.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+//go:build !cgo
+
+package linux
+
+func fqdn() (string, error) {
+	return "", nil
+}
+
+func domainname() (string, error) {
```
Comment (andrewkroh) [inline providers/linux/host_fqdn_nocgo_linux.go:26]: I think the NIS domain can be read from `/proc/sys/kernel/domainname`. Then you wouldn't need separate CGO and non-CGO implementations. And probably the docker based integration testing could be removed since you could simulate various hostname and domainname via the filesystem (probably like  `newLinuxSystem("testdata/my_test_proc")`).

https://www.kernel.org/doc/html/latest/admin-guide/sysctl/kernel.html#domainname-hostname

**Comment by v1v** [comment]
There are test errors, further details in https://github.com/elastic/go-sysinfo/pull/144#issuecomment-1370737951 and https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp%2FPR-144/detail/PR-144/46/tests/

File: providers/darwin/host_fqdn_cgo_darwin.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	domain, err := domainname()
+	if err != nil {
+		return "", err
+	}
+
+	if domain == "" || domain == "(none)" { // mimicking 'hostname -f' behaviour
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply providers/darwin/host_fqdn_cgo_darwin.go:42]: are you sure you haven't set your domain name to "" somehow?
I'm assuming it's your Elastic laptop, thus with Elastic Shield installed. Am I correct?

Have you used the tool in Elastic shield to rename the laptop? That would explain this behaviour

File: providers/darwin/host_fqdn_nocgo_darwin.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+//go:build !cgo
+
+package darwin
+
+func fqdn() (string, error) {
+	return "", nil
+}
+
+func domainname() (string, error) {
+	return "", nil
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply providers/darwin/host_fqdn_nocgo_darwin.go:27]: this is already the no Cgo implementation. I'm already using Cgo for the `getdomainname` syscall.

`sysctl` is an external application, right? I'm avoiding invoking any external application.

File: .ci/scripts/test.sh [hunk 10]
```diff
@@ -37,4 +37,4 @@ fi
 # Run the tests
 export OUT_FILE="build/test-report.out"
 mkdir -p build
-gotestsum --format testname --junitfile "build/junit-${GO_VERSION}.xml" -- ./...
+gotestsum --format testname --junitfile "build/junit-${GO_VERSION}.xml" -- -tags integration -run TestHost_FQDN$ ./...
```
Comment (v1v) [inline .ci/scripts/test.sh:40]: This particular `-tags integration` behaves differently when running on MacOS,

For instance, any builds in the main branch run the tests

image

but this change does not run anything on MacOS

image


I guess that's only for testing purposes, if so, please forgive me :)

but

Comment (AndersonQ) [reply .ci/scripts/test.sh]: that is correct, the tests enabled by the `integration` build flag are only for linux

Comment (v1v) [reply .ci/scripts/test.sh]: If so, then `macos` will need to use a different script, otherwise it will not run any tests

File: providers/darwin/host_fqdn_nocgo_darwin.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrewkroh) [reply providers/darwin/host_fqdn_nocgo_darwin.go:27]: Sorry, I wasn't clear. We can invoke the `sysctl` syscall _without_ needing CGO. Then we can have a single implementation that works for all darwin builds. There should be some examples already in the codebase (if not see the `os.Hostname()` implementation in the stdlib for darwin).

File: providers/linux/host_fqdn_nocgo_linux.go (hunk 7 above)
Comment (leehinman) [reply providers/linux/host_fqdn_nocgo_linux.go]: Do we really want to bring the NIS domain into FQDN?  As I understand it, FQDN is part of DNS not NIS.  [Gnu libc](https://www.gnu.org/software/libc/manual/html_node/Host-Identification.html) and [RFC1594](https://datatracker.ietf.org/doc/html/rfc1594#section-5) seem to agree.

File: providers/shared/fqdn.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+	var errs error
+
+	fqdn, err := fqdnFromEtcHosts(hostname)
+	if err != nil {
+		errs = fmt.Errorf("could not get FQDN, all methods failed: %w", err)
+	}
+	if fqdn != "" {
+		return fqdn, nil
+	}
```
Comment (leehinman) [inline providers/shared/fqdn.go:47]: I think I'm missing something.  Why parse `/etc/hosts`?  The `net.LookupCNAME`, `net.LookuIP` and `net.LookupAddr` will use `/etc/hosts` if the resolver on the system uses that file for name resolution.  Is there a specific use case this is addressing?  If so I really think we need it described in the docstring.

File: providers/linux/host_fqdn_integration_docker_linux_test.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"fmt"
+	"testing"
+)
+
+func TestHost_FQDN_set(t *testing.T) {
+	host, err := newLinuxSystem("").Host()
+	if err != nil {
+		t.Fatal(fmt.Errorf("could not het host information: %w", err))
```
Comment (leehinman) [inline providers/linux/host_fqdn_integration_docker_linux_test.go:30]:
```suggestion
		t.Fatal(fmt.Errorf("could not get host information: %w", err))
```

File: providers/linux/host_fqdn_integration_docker_linux_test.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+	got := host.Info()
+	if got.FQDN != wantFQDN {
+		t.Errorf("got FQDN %q, want: %q", got.FQDN, wantFQDN)
+	}
+}
+
+func TestHost_FQDN_not_set(t *testing.T) {
+	host, err := newLinuxSystem("").Host()
+	if err != nil {
+		t.Fatal(fmt.Errorf("could not het host information: %w", err))
```
Comment (leehinman) [inline providers/linux/host_fqdn_integration_docker_linux_test.go:42]:
```suggestion
		t.Fatal(fmt.Errorf("could not get host information: %w", err))
```

**Review by leehinman — CHANGES_REQUESTED** [review]

File: providers/linux/host_fqdn_nocgo_linux.go (hunk 7 above)
Comment (andrewkroh) [reply providers/linux/host_fqdn_nocgo_linux.go]: I have the same question. I was alluding to this in my other comment at https://github.com/elastic/go-sysinfo/pull/144#discussion_r1088114867. Here I was mainly focused on the code as it was without debating DNS vs NIS. I haven't looked at the latest changes to see how this has evolved, but I still think we need to clear on what we are reporting.

File: go.mod [hunk 14]
```diff
@@ -1,18 +1,35 @@
...
-	golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a
+	github.com/stretchr/testify v1.7.0
+	golang.org/x/sys v0.1.0
 	howett.net/plist v0.0.0-20181124034731-591f970eefbb
 )
 
 require (
-	github.com/davecgh/go-spew v1.1.0 // indirect
-	github.com/pkg/errors v0.8.1 // indirect
+	github.com/Microsoft/go-winio v0.6.0 // indirect
```
Comment (michalpristas) [inline go.mod:16]: this seems off

File: providers/shared/fqdn.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+			continue
+		}
+		return strings.TrimSuffix(names[0], "."), nil
+	}
+
+	return "", errs
+}
+
+func fqdnFromEtcHosts(hostname string) (string, error) {
+	f, err := os.Open(etcHosts)
```
Comment (michalpristas) [inline providers/shared/fqdn.go:74]: not closing file handle

File: providers/shared/fqdn.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+// hostname, if found it returns the canonical_hostname. The canonical_hostname
+// should be the FQDN, see HOSTNAME(1).
+// TODO: check k8s: https://kubernetes.io/docs/tasks/network/customize-hosts-file-for-pods/
+func findInHostsLine(hostname, hostsEntry string) string {
+	line, _, _ := strings.Cut(hostsEntry, "#")
+	if len(line) < 1 {
+		return ""
+	}
+
+	fileds := strings.FieldsFunc(line, func(r rune) bool {
```
Comment (michalpristas) [inline providers/shared/fqdn.go:119]: s/fileds/fields

File: providers/shared/fqdn.go [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+	if len(fileds) < 2 {
+		// invalid hostsEntry
+		return ""
+	}
+
+	// fields[0] is the ip address
+	cannonical, aliases := fileds[1], fileds[1:]
+
+	// TODO: confirm: a name should not repeat on different addresses.
```
Comment (michalpristas) [inline providers/shared/fqdn.go:131]: adressed or issue please

File: providers/windows/host_windows.go [hunk 18]
```diff
@@ -137,6 +143,49 @@ func (r *reader) hostname(h *host) {
...
+		// https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getcomputernameexw#return-value
+		if errors.Is(err, syscall.ERROR_MORE_DATA) {
+			if size <= uint32(len(buff)) {
+				// Safeguard to avoid an infinite loop.
+				return "", fmt.Errorf(
+					"windows.GetComputerNameEx returned ERROR_MORE_DATA, " +
+						"but data size should fit into buffer")
+			} else {
+				// Grow the buffer and try again.
+				// Should we limit its growth?
```
Comment (michalpristas) [inline providers/windows/host_windows.go:179]: yes

**Comment by jlind23** [comment]
@AndersonQ would it also be possible to update the description with the algorithm you used when default values were used, etc..

**Comment by AndersonQ** [comment]
> @AndersonQ would it also be possible to update the description with the algorithm you used when default values were used, etc..

@jlind23 done

File: providers/windows/host_windows.go (hunk 18 above)
Comment (AndersonQ) [reply providers/windows/host_windows.go]: but then what would be the limit?
`size` is the needed size for a buffer to fit the FQDN. The unix implementation does not limit the size, so I'd say it should be fine. it hardly wouldn't fit into memory, if it don't, I'd say windows is on a lot more trouble than us.

File: go.mod (hunk 14 above)
Comment (AndersonQ) [reply go.mod:16]: besides the only direct dependency added was `github.com/docker/docker`

**Comment by AndersonQ** [comment]
@michalpristas, @andrewkroh, @leehinman, @fearful-symmetry I updated the PR, it's ready for final review. I still need to add the changes to the changelog/readme though

File: providers/linux/host_fqdn_integration_linux_test.go [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+const wantFQDN = wantHostname + "." + wantDomain
+
+func TestHost_FQDN(t *testing.T) {
+	const envKey = "GO_VERSION"
+	goversion, ok := os.LookupEnv(envKey)
+	if !ok {
+		t.Fatalf("environment variable %s not set, please set a Go version",
+			envKey)
+	}
+	image := "golang:" + goversion
```
Comment (fearful-symmetry) [inline providers/linux/host_fqdn_integration_linux_test.go:46]: Extremely pedantic question: What, exactly, are we trying to test with the containers? Just how this will behave in a containerized environment in general? If we wanted to test this under different OS-specific conditions, we might want to target containers built around centos, Feodra, ubuntu, etc.

File: providers/shared/fqdn.go (hunk 11 above)
Comment (leehinman) [reply providers/shared/fqdn.go:47]: yep they go through '/etc/hosts'  the pure go one use [readHost](https://github.com/golang/go/blob/e2b620ce730992c7991951b6b2d005f70e8c8d36/src/net/hosts.go#L50) where the `testHookHostsPath` is `/etc/hosts` and the cgo implementations all use the libc resolver which will use `/etc/hosts` if your system is setup to do that, and if it isn't setup that way you don't want to parse `/etc/hosts` anyway.

**Review by leehinman — CHANGES_REQUESTED** [review]

File: providers/darwin/host_darwin_test.go [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+func TestHost(t *testing.T) {
+	host, err := darwinSystem{}.Host()
+	if err != nil {
+		t.Logf("could not get all host info: %v\n", err)
+	}
+
+	info := host.Info()
+	data, _ := json.MarshalIndent(info, "", "  ")
+	t.Logf(string(data))
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply providers/darwin/host_darwin_test.go:39]: No it does not. I just followed the pattern that was already in place for [linux](https://github.com/elastic/go-sysinfo/blob/main/providers/linux/host_linux_test.go#L32-L41).

File: providers/linux/host_fqdn_integration_linux_test.go (hunk 19 above)
Comment (AndersonQ) [reply providers/linux/host_fqdn_integration_linux_test.go:46]: yes that is correct. I haven't found any testing checking hostname info. That was the easiest way I found to set up an environment I can control hostname and domain to test. It isn't comprehensive, but it's more that there was before.

**Review by leehinman — APPROVED** [review]
LGTM
